### ISSUE TITLE
Zarr V3: questions about fill_value representation as hexadecimal representation and complex data types

### ISSUE THREAD

**Issue by rouault** [zarr-developers/zarr-specs#231]
Reading https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#fill-value, I'm not sure how the nodata value for complex data types should be encoded. Should it be an array of two elements like ``[1.5, 2.5]`` ? Is an hexadecimal representation allowed, and if so how should it be encoded ?

Regarding the hexadecimal representation, it is a bit ambiguous for Float data types. I assume this should be the integer corresponding to the IEEE754 encoding of the floating point number, and that for a Float32 it should be 0xABCDEFGH where AB is the most significant byte ?

**Comment by jstriebel** [comment]
Agreed, this is currently both missing. @rouault or @jbms, would one of you submit a PR?

**Comment by rouault** [comment]
> would one of you submit a PR?

Probably a bit beyond what I can properly write in proper English in a formal doc :-) I'm not clear either how you would format "two consecutive little endian IEEE754 binary64 values" in a string.  Probably would require some separator, like "0xEFCDAB8967452301,0xEFCDAB8967452301" but that's going to be painful to parse, and I believe that requiring an array of strings or array of numbers for complex data is better.

And I'm not sure why "little endian" is mentionned. An hexadecimal representation usually uses a MSB convention (like the decimal writing), independently of the machine byte ordering. e.g "0xEFCDAB8967452301" represents the 64-bit word of least significant part 01 and most significant part 0xEF.
That's also how Python / C library formats numbers. For example the representation of 1.5 as the hexadecimal representation of the corresponding IEEE754 64-bit float would be:
>>> hex(struct.unpack('Q', struct.pack('d', 1.5))[0])
'0x3ff8000000000000'

**Comment by jstriebel** [comment]
> "two consecutive little endian IEEE754 binary64 values" in a string

I assume that we would simply represent them as a JSON array in this case instead of a string.

**Comment by jstriebel** [comment]
> And I'm not sure why "little endian" is mentionned. An hexadecimal representation usually uses a MSB convention (like the decimal writing), independently of the machine byte ordering. e.g "0xEFCDAB8967452301" represents the 64-bit word of least significant part 01 and most significant part 0xEF.

I think that only holds true for little-endian representations. A hexadecimal representation usually follows the memory layout, e.g. the first byte is represented left in the hexadecimal representation, and therefore depends if the float is interpreted as little-endian or big-endian.

That's also how Python / C library formats numbers. For example the representation of 1.5 as the hexadecimal representation of the corresponding IEEE754 64-bit float would be:
> 
> ```
> >>> hex(struct.unpack('Q', struct.pack('d', 1.5))[0])
> '0x3ff8000000000000'
> ```

There is a difference for little and big endian float64 representations:
```python
>>> hex(struct.unpack('Q', struct.pack('<d', 1.5))[0])
'0x3ff8000000000000'
>>> hex(struct.unpack('Q', struct.pack('>d', 1.5))[0])
'0xf83f'  # the first six bytes are just zeros
```

**Comment by rouault** [comment]
> There is a difference for little and big endian float64 representations:
```
>>> hex(struct.unpack('Q', struct.pack('>d', 1.5))[0])
'0xf83f'  # the first six bytes are just zeros
```

The above is "unfair". If you force big endian order for the floating-point representation, you also need to force it for the integer one, and they both cancel:
```
>>> hex(struct.unpack('>Q', struct.pack('>d', 1.5))[0])
'0x3ff8000000000000'
```

What I mean is that the usual/canonical way to express the hexadecimal representation of a number is to have the most significants bytes/bits at the beginning, so it is more like a big-endian representation than a little-endian one. But I don't think that speaking about endianness here is appropriate. I'm not sure about the appropriate terminology

**Comment by jbms** [comment]
In the PR I created I proposed removing this byte representation altogether.

But as far as big endian vs little endian, yes for a single floating point number I agree that big endian is the more natural text representation.  But I think the idea behind this representation was for it to work for any data type, and so it needs to be specified in a generic way, like take the default binary representation and convert to a hex string.  Given that we use little endian by default in the chunk encoding, it would be odd to say that the fill value uses big endian representation.  It would also be odd to just use the default binary encoding and then reverse the byte order, since then complex numbers would have the imaginary component first.
